packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+      }
+      return null;
+    });
+  }
+}
+
+// TODO: We may want to remodel dispatch to either extend ChainRpcProvider or just move all the ChainRpcProvider functionality over here.
+// For now, a sep. class for simplicity of introducing new functionality.
+/**
+ * @classdesc Dispatches and monitors transaction queue.
+ */
+export class TransactionDispatch {
+  public readonly chainId: number;
+
+  private readonly submitQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  // A queue for execution of transaction validation. When it completes, we'll shift the transaction from
+  // pending stack -> validated stack in the buffer.
+  private readonly validateQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  private readonly buffer: TransactionBuffer = new TransactionBuffer();
+
+  /**
+   * Centralized transaction dispatch management class.
+   *
+   * @param logger The pino.BaseLogger instance we use for logging.
+   * @param provider The ChainRpcProvider instance we use for interfacing with the chain.
+   */
+  constructor(private readonly logger: BaseLogger, private readonly provider: ChainRpcProvider) {
+    this.chainId = this.provider.chainId;
+  }
+
+  public async submit(minTx: WriteTransaction): Promise<Transaction> {
+    const method = this.submit.name;
+    // get gas estimate
+    let gasLimit: BigNumber;
+    let result = await this.provider.estimateGas(minTx);
+    if (result.isErr()) {
+      if (result.error instanceof TransactionReverted) {
+        // If we get a TransactionReverted error, that means the gas estimate call
+        // indicated our transaction would fail on-chain. The details of the failure will
+        // be included in the error.
+        throw result.error;
+      }
+      this.logger.warn(
+        {
+          method,
+          transaction: minTx,
+          error: result.error,
+        },
+        "Estimate gas failed due to an unexpected error.",
+      );
+      throw result.error;
+    } else {
+      gasLimit = result.value;
+    }
+
+    // get gas price
+    result = await this.provider.getGasPrice();
+    if (result.isErr()) {
+      throw result.error;
+    }
+    const gasPrice = new GasPrice(result.value, gasLimit);
+
+    // Queue up the transaction with these values.
+    const submitResult = await this.submitQueue.add(
+      async (): Promise<{ value: Transaction | Error; success: boolean }> => {
+        // NOTE: This call must be here, serialized within the queue, as it is depenedent on pending transaction count.
+        // If transaction buffer returns null, that indicates the buffer is empty; meaning we haven't sent any transactions yet.
+        let nonce = await this.buffer.getNonce();
+        if (!nonce) {
+          const result = await this.provider.getTransactionCount();
+          if (result.isErr()) {
+            throw result.error;
+          } else {
+            nonce = result.value;
+          }
+        }
```

Comment:
This *entire* process should be handled within the buffer imo -- we should always only need to look 1 place for the logic around handling nonces

Reply:
Sure, agreed. would be cleaner

Reply:
But should only ever be called here